src/infra/core/src/compact_service_impl.rs

```diff
@@ -0,0 +1,378 @@
+// Copyright Kamu Data, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use std::fs;
+use std::path::{Path, PathBuf};
+use std::sync::Arc;
+
+use datafusion::prelude::*;
+use dill::{component, interface};
+use domain::compact_service::{
+    CompactError,
+    CompactionMultiListener,
+    CompactionPhase,
+    InvalidDatasetKindError,
+    NullCompactionListener,
+};
+use futures::stream::TryStreamExt;
+use kamu_core::compact_service::CompactService;
+use kamu_core::*;
+use opendatafabric::{
+    AddData,
+    Checkpoint,
+    DatasetHandle,
+    DatasetKind,
+    MetadataEvent,
+    Multihash,
+    OffsetInterval,
+    SourceState,
+};
+use random_names::get_random_name;
+use url::Url;
+
+use crate::*;
+
+pub struct CompactServiceImpl {
+    dataset_repo: Arc<dyn DatasetRepository>,
+    dataset_authorizer: Arc<dyn domain::auth::DatasetActionAuthorizer>,
+    time_source: Arc<dyn SystemTimeSource>,
+    run_info_dir: PathBuf,
+}
+
+#[derive(Debug, Default, Clone)]
+struct DataSliceBatchInfo {
```

Comment:
Suggestions:
1) Make an enum to distinguish compacted vs non-compacted batch:

```
enum DataSliceBatch {
    CompactedBatch( DataSliceBatchInfo),
    SingleBlock(...)
}
```
This should reduce number of `Option` fields.

2) Simplify how `prev_*` and `new_*` vars are handled by grouping into child structs:
```

#[derive(Debug, Default, Clone)]
struct DataSliceBatchUpperBound {
    pub new_source_state: Option<SourceState>,
    pub new_checkpoint: Option<Checkpoint>,    
    pub new_end_offset: Option<u64>,
}

#[derive(Debug, Default, Clone)]
struct DataSliceBatchLowerBound {
    pub prev_offset: Option<u64>,
    pub prev_checkpoint: Option<Multihash>,
    pub prev_end_offset: Option<u64>,
}
```